crates/rattler_index/src/lib.rs

```diff
@@ -229,53 +269,80 @@ async fn index_subdir(
-                        let reader = buffer.reader();
-                        // We already know it's not None
-                        let archive_type = ArchiveType::try_from(&filename).unwrap();
-                        let record = match archive_type {
-                            ArchiveType::TarBz2 => package_record_from_tar_bz2_reader(reader),
-                            ArchiveType::Conda => package_record_from_conda_reader(reader),
-                        }?;
-                        pb.inc(1);
-                        Ok::<(String, PackageRecord), std::io::Error>((filename.clone(), record))
-                    }
+    let mut tasks = FuturesUnordered::new();
+    for filename in packages_to_add.iter() {
+        let task = {
+            let op = op.clone();
+            let filename = filename.clone();
+            let pb = pb.clone();
+            let semaphore = semaphore.clone();
+            {
+                async move {
+                    let _permit = semaphore
+                        .acquire()
+                        .await
+                        .expect("Semaphore was unexpectedly closed");
+                    pb.set_message(format!(
+                        "Indexing {} {}",
+                        subdir.as_str(),
+                        console::style(filename.clone()).dim()
+                    ));
+                    let file_path = format!("{subdir}/{filename}");
+                    let buffer = op.read(&file_path).await?;
+                    let reader = buffer.reader();
+                    // We already know it's not None
+                    let archive_type = ArchiveType::try_from(&filename).unwrap();
+                    let record = match archive_type {
+                        ArchiveType::TarBz2 => package_record_from_tar_bz2_reader(reader),
+                        ArchiveType::Conda => package_record_from_conda_reader(reader),
+                    }?;
+                    pb.inc(1);
+                    // todo: make this future ok/err instead of results
+                    Ok::<(String, PackageRecord), std::io::Error>((filename.clone(), record))
                 }
-            })
-        })
-        .collect::<Vec<_>>();
-    let results = try_join_all(tasks).await?;
-
-    pb.finish_with_message(format!("Finished {}", subdir.as_str()));
+            }
+        };
+        tasks.push(tokio::spawn(task));
+    }
+    let mut results = Vec::new();
+    while let Some(join_result) = tasks.next().await {
+        match join_result {
+            Ok(Ok(result)) => results.push(result),
+            Ok(Err(e)) => {
+                tasks.clear();
+                tracing::error!("Failed to process package: {}", e);
+                pb.abandon_with_message(format!(
+                    "{} {}",
+                    console::style("Failed to index").red(),
+                    console::style(subdir.as_str()).dim()
+                ));
```

Comment:
this is not perfect as the progress bar can get overwritten by other tasks that are still executed after this failure. haven't seen a nice way to avoid this behavior

<img width="1351" alt="image" src="https://github.com/user-attachments/assets/4aa61039-bcb5-4d3f-ab0f-42932aa0ee8e" />

i think we can keep it for now like this since the error messages are printed anyway and i expect this script to be executed in non-interactive environments anyway most of the time